src/execution_plan/ops/shared/update_functions.c

```diff
@@ -0,0 +1,147 @@
+			Schema_AddNodeToIndices(s, (Node *)update->ge);
+		}
+	}
+
+	if(stats) stats->properties_set += properties_set;
+}
+
+void EvalEntityUpdates(GraphContext *gc, PendingUpdateCtx **updates, Record r, char *alias,
+					   EntityUpdateEvalCtx *ctx, bool allow_null) {
+	Schema *s         = NULL;
+	int label_id      = GRAPH_NO_LABEL;
+	const char *label = NULL;
+	bool node_update  = false;
+
+	// Get the type of the entity to update. If the expected entity was not
+	// found, make no updates but do not error.
+	RecordEntryType t = Record_GetType(r, ctx->record_idx);
+	if(t == REC_TYPE_UNKNOWN) return;
+
+	// Make sure we're updating either a node or an edge.
+	if(t != REC_TYPE_NODE && t != REC_TYPE_EDGE) {
+		ErrorCtx_RaiseRuntimeException("Update error: alias '%s' did not resolve to a graph entity",
+									   alias);
+	}
+
+	GraphEntity *entity = Record_GetGraphEntity(r, ctx->record_idx);
+
+	if(t == REC_TYPE_NODE) node_update = true;
+
+	// If the entity is a node, set its label if possible.
+	if(node_update) {
+		Node *n = (Node *)entity;
+		// Retrieve the node's local label if present.
+		label = NODE_GET_LABEL(n);
+		// Retrieve the node's Label ID from a local member or the graph.
+		label_id = NODE_GET_LABEL_ID(n, gc->g);
+		if((label == NULL) && (label_id != GRAPH_NO_LABEL)) {
+			// Label ID has been found but its name is unknown, retrieve its name and update the node.
+			s = GraphContext_GetSchemaByID(gc, label_id, SCHEMA_NODE);
+			label = Schema_GetName(s);
+			n->label = label;
+		}
+	}
+
+	// If this update replaces all existing properties, enqueue an update to do so.
+	if(ctx->mode == UPDATE_REPLACE) {
+		PendingUpdateCtx update = {
+			.ge            =  entity,
+			.label_id      =  label_id,
+			.update_index  = (label_id != GRAPH_NO_LABEL),
+			.attr_id       =  ATTRIBUTE_ALL,
+		};
+		*updates = array_append(*updates, update);
+	}
+
+	uint exp_count = array_len(ctx->properties);
+	for(uint i = 0; i < exp_count; i++) {
+		bool update_index = false;
+		SIValue new_value = AR_EXP_Evaluate(ctx->properties[i].value, r);
+
+		SIType accepted_properties = SI_VALID_PROPERTY_VALUE;
+		// If we're converting a SET clause, NULL is acceptable as it indicates a deletion.
+		if(allow_null) accepted_properties |= T_NULL;
+		// Emit an error and exit if we're trying to add an invalid type.
+		if(!(SI_TYPE(new_value) & accepted_properties)) {
+			Error_InvalidPropertyValue();
+			ErrorCtx_RaiseRuntimeException(NULL);
+			break;
+		}
+
+		Attribute_ID attr_id = ctx->properties[i].id;
+		// Determine whether we must update the index for this update.
+		if(node_update && label_id != GRAPH_NO_LABEL) {
+			// If the (label:attribute) combination has an index, take note.
+			update_index = GraphContext_GetIndex(gc, label, &attr_id, IDX_ANY) != NULL;
+		}
+
+		PendingUpdateCtx update = {
+			.ge            =  entity,
+			.label_id      =  label_id,
+			.new_value     =  new_value,
+			.update_index  =  update_index,
```

Comment:
This `update` describes a single update to an entity, make sure that if a specific entity has a number of attributes updated say `SET n.a = 1, n.b = 1` where both `a` and `b` are indexed fields then `n` must be re-indexed only once, otherwise this will hurt performance.